pkg/controller/volume/pvcprotection/pvc_protection_controller.go

```diff
@@ -275,6 +282,27 @@ func (c *Controller) askAPIServer(pvc *v1.PersistentVolumeClaim) (bool, error) {
 }
 
 func podUsesPVC(pod *v1.Pod, pvc string) bool {
+	// A pod that has a deletionTimestamp and a zero
+	// deletionGracePeriodSeconds
+	// a) has been processed by kubelet and is ready for deletion or
```

Comment:
This is not accurate. deleteGracePeriodSeconds tells kubelet when it can start force terminating the containers. Volume teardown only starts after containers are termianted. So there is an additional time period after the grace period where volume teardown is happening.

And the main signal we have today that volume teardown has finished is the deletion of the Pod object.

Adding this check to removing the finalizer early will introduce races where the volume gets deleted before unmount is complete, potentially causing unmount to hang/fail forever.

We need to think of a different way to get the signal that unmount is finished. And potentially we may still have a race for plugins that support unstage because the logic that kubelet uses to check if the pod resources are cleaned up only checks for pod volume directories, not plugin directories. cc @jsafrane

Reply:
I have pushed an updated version where the PVC protection controller change is under a feature gate check. Without the feature, it should behave as before.

I also added a TODO about this questionable check to the code and my tracking list: https://github.com/kubernetes/enhancements/issues/1698#issuecomment-655344680

Reply:
I also placed the extension of the PodPVCIndexer under the feature check because `podUsesPVC` is expecting the traditional behavior when the feature is off.